### DISCUSSION TITLE
单看视频里的原理解释，这种扫描算法是不是得配个校准程序？

### DISCUSSION THREAD

**Discussion by Mehver** [peng-zhihui/HelloWord-Keyboard#35] [Q&A] [answered]
让我自己分析这算法，属于是自取其辱了。所以有没有dalao能解释解释？

当然如果1:1复刻瀚文的话，每个按键对应的时序就不用调了。

但是如果想自己画PCB，应该得逐个按键去调时序吧？至少改变键盘配列之后肯定要改软件部分。

这是我的理解，如果到这里都没错的话，是不是可以做个可视化的校准程序？类似“键盘触发测试”那种，让使用者按指示按下要校对的按键。

之所以有这个想法，也是因为单看视频我感觉这种误差和老化后可能会造成时序的偏移。视频里说道精度似乎很高，但还是想到了这，于是就问问。求指点

**Comment by xingrz** [comment, marked as answer]
<img width="2322" alt="image" src="https://user-images.githubusercontent.com/288288/183246969-ae954ae7-09af-4640-8189-01ba67c80881.png">

这个算法其实十分好理解。STM32的SPI1_MISO接在了U3这第一颗74HC165的串行输出上，而第二颗（U4）的串行输出又接在了第一颗的串行输入上。以此类推，总共有11颗74HC165。

用大白话说，软件上从SPI1一口气读取11个字节，那么第1个字节就是U3，第2个字节就是U4…第11个字节是U13。而每个字节，第1位对应这颗74HC165的D0，第2位对应D1…第8位对应D7。话虽如此，如果大幅度改动了按键顺序的话，重新对位也是个不小的工作量。

不知道你所值的时序偏移是什么意思（抱歉，我入行不久，还没足够的时间让我遇到老化问题【），是指时钟误差？。但是我的理解，74HC165的输出可以看作是SPI协议，它是依靠从MCU往CP脚输出的时钟信号（的上升沿）给出对应输出的。所以快点或者慢点都没关系。

**Reply by pthuang** [reply]
spi的时钟是主机给的，这里是单片机（给出的时钟）主动扫描，时钟应该是内部定时器产生的，不大可能偏移，设备老化确实会导致时钟精度下降，但这个时间也是几年后了，而且最多偏个ns级别（可能都不到），以8M的晶振为例，周期125ns，偏移几乎可以忽略。而且spi是边沿采样，如果严格按照时钟边沿和数据最中间对齐，就算偏一点数据也不会采错，所以不存在你说的问题。时序是不需要校准的，但是如果改pcb比如扩展按键肯定需要扩展电路加移位寄存器，然后固件里的键盘映射肯定是要扩展的。

**Reply by pthuang** [reply]
1ppm的8M晶振出厂时钟偏移在正负8Hz，换算成周期连正负1ps都不到，老化后估计也是ps级别的偏移，等它偏移太大的时候，你这键盘指不定在哪吃灰呢。

**Reply by pthuang** [reply]
xikii不好意思，我好像回复错人了

**Reply by xingrz** [reply]
> xikii不好意思，我好像回复错人了

没事，我也受益匪浅

**Comment by peng-zhihui** [comment]
不需要调时序，就是标准的边沿采样不会有任何误差。这个方案layout也随便走线，所有键位映射都在代码里调整，ReMap函数的第0层就是用来映射实际按键走线布局和标准布局的，后面的1、2、3、4...层就是映射从标准布局到自定义布局，可以去看看代码

**Comment by Mehver** [comment]
@xingrz @pthuang @peng-zhihui

感谢讲解，明白了
